examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java

```diff
@@ -16,182 +16,90 @@
+        try {
+            if (args.length != 3) {
+                Utils.printHelp("This example takes 3 parameters (i.e. 6 3 10000):%n" +
+                    "- partition: number of partitions for input and output topics (required)%n" +
+                    "- instances: number of application instances (required)%n" +
+                    "- records: total number of records (required)");
+                return;
             }
-            Thread.sleep(1000);
-        }
-
-        // Create topics in a retry loop
-        while (true) {
-            final short replicationFactor = 1;
-            final List<NewTopic> newTopics = Arrays.asList(
-                new NewTopic(INPUT_TOPIC, numPartitions, replicationFactor),
-                new NewTopic(OUTPUT_TOPIC, numPartitions, replicationFactor));
-            try {
-                adminClient.createTopics(newTopics).all().get();
-                System.out.println("Created new topics: " + newTopics);
-                break;
-            } catch (ExecutionException e) {
-                if (!(e.getCause() instanceof TopicExistsException)) {
-                    throw e;
-                }
-                System.out.println("Metadata of the old topics are not cleared yet...");
 
-                deleteTopic(adminClient, topicsToDelete);
+            int numPartitions = Integer.parseInt(args[0]);
+            int numInstances = Integer.parseInt(args[1]);
+            int numRecords = Integer.parseInt(args[2]);
+
+            // stage 1: clean any topics left from previous runs
+            Utils.recreateTopics(BOOTSTRAP_SERVERS, numPartitions, INPUT_TOPIC, OUTPUT_TOPIC);
+
+            // stage 2: send demo records to the input-topic
+            CountDownLatch producerLatch = new CountDownLatch(1);
+            Producer producerThread = new Producer(
+                "producer", BOOTSTRAP_SERVERS, INPUT_TOPIC, false, null, true, numRecords, -1, producerLatch);
+            producerThread.start();
+            if (!producerLatch.await(2, TimeUnit.MINUTES)) {
+                Utils.printErr("Timeout after 2 minutes waiting for data load");
+                producerThread.shutdown();
+                return;
+            }
 
-                Thread.sleep(1000);
+            // stage 3: read from input-topic, process once and write to the output-topic
+            CountDownLatch processorsLatch = new CountDownLatch(numInstances);
+            for (int instanceIdx = 0; instanceIdx < numInstances; instanceIdx++) {
+                ExactlyOnceMessageProcessor messageProcessor = new ExactlyOnceMessageProcessor(
+                    INPUT_TOPIC, OUTPUT_TOPIC, instanceIdx, processorsLatch);
+                messageProcessor.start();
+            }
+            if (!processorsLatch.await(2, TimeUnit.MINUTES)) {
+                Utils.printErr("Timeout after 2 minutes waiting for record copy");
+                //processors.forEach(ExactlyOnceMessageProcessor::shutdown);
```

Comment:
why commented out?

Reply:
There was no shutdown before the new processor merge. Added now.